utils.py

```diff
@@ -34,3 +35,33 @@ def get_spotify_token(request):
     else:
         # Error getting saved token
         return None
+
+def format_audio_features(track_features):
+    """
+    Optimized function to format the Spotify audio features for model prediction.
+
+    Parameters:
+    - track_features: The audio features of the track as obtained from the Spotify API.
+
+    Returns:
+    - A formatted pandas DataFrame suitable for making predictions with the loaded model.
+    """
+    # Directly create the DataFrame with the desired column order and renaming
+    ordered_features = [
+        ('popularity', 1),  # Assuming popularity is a static placeholder
+        ('duration_ms', 'length'),  # Rename 'duration_ms' to 'length'
+        'danceability', 'acousticness', 'energy', 'instrumentalness',
+        'liveness', 'valence', 'loudness', 'speechiness', 'tempo',
+        'key', 'time_signature'
+    ]
+
+    # Prepare a dictionary for renaming and column order
+    rename_dict = {orig: new if isinstance(new, str) else orig
+                   for orig, new in ordered_features}
+    columns_order = [new if isinstance(new, str) else orig
+                     for orig, new in ordered_features]
+
+    # Select, rename, and reorder columns in one go
+    spotify_data = pd.DataFrame(track_features).rename(columns=rename_dict)[columns_order]
+
```

Comment:
track_features is missing popularity data

Reply:
The reason popularity data here is ignored, is because spotify's /audio-features API [no longer offer popularity feature](https://developer.spotify.com/documentation/web-api/reference/get-audio-features).

And stripping popularity data from training data doesn't seem to impact the classifier's accuracy a lot (intuitively, this make sense, right?):

With popularity feature in training data:
```
Accuracy for Gradient Boosting Classifier : 0.848
Classification Report for Gradient Boosting Classifier : 
              precision    recall  f1-score   support

       Happy       0.70      0.62      0.65        26
         Sad       0.95      0.91      0.93        44
   Energetic       0.70      0.81      0.75        32
        Calm       0.97      0.97      0.97        36

    accuracy                           0.85       138
   macro avg       0.83      0.83      0.83       138
weighted avg       0.85      0.85      0.85       138
```

Without popularity feature in training data:
```
Accuracy for Gradient Boosting Classifier : 0.841
Classification Report for Gradient Boosting Classifier : 
              precision    recall  f1-score   support

       Happy       0.70      0.59      0.64        27
         Sad       0.93      0.91      0.92        43
   Energetic       0.73      0.82      0.77        33
        Calm       0.94      0.97      0.96        35

    accuracy                           0.84       138
   macro avg       0.82      0.82      0.82       138
weighted avg       0.84      0.84      0.84       138
```

The tradeoff of stripping the popularity feature here is marginal, so I recommend just ignoring it in this case.